reference/src/glossary.md

```diff
@@ -149,16 +149,23 @@ For some more information, see [this blog post](https://www.ralfj.de/blog/2018/0
 #### Undefined Behavior
 [ub]: #undefined-behavior
 
-*Undefined Behavior* is a concept of the contract between the Rust programmer and the compiler:
-The programmer promises that the code exhibits no undefined behavior.
-In return, the compiler promises to compile the code in a way that the final program does on the real hardware what the source program does according to the Rust Abstract Machine.
-If it turns out the program *does* have undefined behavior, the contract is void, and the program produced by the compiler is essentially garbage (in particular, it is not bound by any specification; the program does not even have to be well-formed executable code).
-
-In Rust, the [Nomicon](https://doc.rust-lang.org/nomicon/what-unsafe-does.html) and the [Reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) both have a list of behavior that the language considers undefined.
-Rust promises that safe code cannot cause Undefined Behavior---the compiler and authors of unsafe code takes the burden of this contract on themselves.
-For unsafe code, however, the burden is still on the programmer.
-
-Also see: [Soundness][soundness].
+*Undefined Behavior* (UB) is the behavior for which the Rust language provides no guarantees. Rust programs that do **not** exhibit *undefined behavior* execute according to the semantics of the Rust Abstract Machine on real hardware.
+Rust imposes no requirements on programs that exhibit undefined behavior; these programs are not bound by any specification.
```

Comment:
So, I like your start, because it is more concrete than mine. However, I think it is very important that the term "contract" comes up in the first paragraph already. That is the mindset needed to understand UB, it should not be an afterthought degraded to a "note".
Also, your text starts by explaining things about the negation, which is confusing.

But I realized that I am not happy how the Abstract Machine comes in in either version -- after all, it is the abstract machine that *defines* UB; the text makes it sound like the Abstract Machine only enters the picture when there is no UB.

So here's another shot, merging your and my versions:

> *Undefined Behavior* (UB) is an error condition of the Rust Abstract Machine.
> The Rust language provides no guarantees for programs that can reach this error condition.
> As part of the contract between the Rust programmer and the compiler, is thus the obligation of the programmer to ensure that their code, when run according to the rules of the Rust Abstract Machine, does not exhibit undefined behavior.
> In return, the compiler guarantees that programs that do **not** exhibit undefined behavior execute on real hardware according to the semantics of the Rust Abstract Machine.
> Rust imposes no requirements on the behavior of programs that exhibit undefined behavior; these programs are not bound by any specification (in particular, the program produced by the compiler is essentially garbage, it does not even have to be well-formed executable code).